bin/add_frontend_server.rb

```diff
@@ -67,41 +83,72 @@ def aws_instance_type(environment)
+        'Name' && tag.value.include?('frontend') } }
   end
 
   frontend_instances.flatten!
 
-  instance_distribution = frontend_instances.each_with_object(Hash.new(0)){|(instance, _), instance_distribution|
-    instance_distribution[instance.placement.availability_zone] += 1}
-  determined_instance_zone, instance_count = instance_distribution.min_by{|_, v| v}
+  instance_distribution = frontend_instances.each_with_object(Hash.new(0)) { |(instance, _), instance_distribution|
+    instance_distribution[instance.placement.availability_zone] += 1 }
+
+  instance_distribution
+end
+
+# Return an array of objects that have the names and zones of instances to create
+#
+# When we are creating multiple frontends, don't bother trying to distribute in the underscaled instances. Just
+# distribute them across all of our existing zones
+def create_instance_provisioning_infos
+  instances_to_provision_array = []
+  instance_distribution = determine_frontend_instance_distribution
+  instances_to_provision = @options['count'].to_f
+
+  instances_to_provision_for_zones = Hash.new()
+
+  instance_distribution.keys.each do |zone|
+    instances_to_provision_for_zones[zone] = (instances_to_provision / instance_distribution.keys.count).floor
+  end
+
+  leftover_instances = instances_to_provision % instance_distribution.keys.count
+
+  instance_distribution.keys.each do |zone|
+    break if leftover_instances == 0
+    instances_to_provision_for_zones[zone] += 1
+    leftover_instances -= 1
+  end
 
-  puts "Using underscaled instance zone #{determined_instance_zone}"
+  puts instances_to_provision_for_zones
 
+  instances_to_provision_for_zones.each do |zone, count|
+    count.times do |new_instances|
+      new_instance_name = determine_unique_name_for_instance_zone(@username, nil, zone, new_instances + instance_distribution[zone])
+      instances_to_provision_array << InstanceProvisioningInfo.new(zone, new_instance_name)
+    end
+  end
+
+  instances_to_provision_array
+end
+
+# Get a unique name for the instance zone. Make sure there isn't already a chief client or node with this name
+def determine_unique_name_for_instance_zone(ssh_username, frontend_name, determined_instance_zone, instance_count)
+  puts "Using instance zone #{determined_instance_zone} which has #{instance_count} instances"
   # There may be multiple instances of this script running at once or
   # stale clients or nodes in Chef, so try a few different instance indices to
   # find a unique one.
+
   Net::SSH.start('gateway.code.org', ssh_username) do |ssh|
     retry_index = 0
     while retry_index < 50
       if frontend_name
         name = "#{frontend_name}#{retry_index == 0 ? '' : retry_index.to_s}"
       else
-        name = "frontend-#{determined_instance_zone[-1, 1] + (instance_count + 1 + retry_index).to_s}"
+        name = @options['prefix'] + "frontend-#{determined_instance_zone[-1, 1] + (instance_count + 1 + retry_index).to_s}"
```

Comment:
Could you make this a "midfix" instead of a prefix.  I like all of the frontend names to begin with "frontend" for easy for search. 

So if the provided string is hoc, I'd like to see names like "frontend-hoc-*" .